system-contracts/contracts/erc4337/EntryPointV01.sol

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {IEntryPoint, PackedUserOperation} from "../interfaces/IEntryPoint.sol";
+import {IBootloaderUtilities} from "../interfaces/IBootloaderUtilities.sol";
+import {IAccount} from "../interfaces/IAccount.sol";
+import {IAccountCodeStorage} from "../interfaces/IAccountCodeStorage.sol";
+import {INonceHolder} from "../interfaces/INonceHolder.sol";
+import {IContractDeployer} from "../interfaces/IContractDeployer.sol";
+import {ISystemContext} from "../interfaces/ISystemContext.sol";
+import {ContractDeployer} from "../ContractDeployer.sol";
+import {Transaction, TransactionHelper, EIP_712_TX_TYPE, LEGACY_TX_TYPE, EIP_2930_TX_TYPE, EIP_1559_TX_TYPE} from "../libraries/TransactionHelper.sol";
+import {BOOTLOADER_FORMAL_ADDRESS, SYSTEM_CONTEXT_CONTRACT, NONCE_HOLDER_SYSTEM_CONTRACT, ACCOUNT_CODE_STORAGE_SYSTEM_CONTRACT, DEPLOYER_SYSTEM_CONTRACT} from "../Constants.sol";
+import {SystemContractBase} from "../abstract/SystemContractBase.sol";
+import {SystemContext} from "../SystemContext.sol";
+import {RLPEncoder} from "../libraries/RLPEncoder.sol";
+import {EfficientCall} from "../libraries/EfficientCall.sol";
+import {UnsupportedTxType, InvalidSig, SigField, HashMismatch} from "../SystemContractErrors.sol";
+
+import {SystemContractsCaller, CalldataForwardingMode} from "../libraries/SystemContractsCaller.sol";
+
+/**
+ * @author Matter Labs
+ * @custom:security-contact security@matterlabs.dev
+ * @notice An EIP4337 EntryPoint contract implementation
+ * built on top of the native ZKsync Account Abstraction.
+ */
+contract EntryPointV01 is IEntryPoint, SystemContractBase {
+    using TransactionHelper for *;
+
+    function handleUserOps(PackedUserOperation[] calldata _ops) external {
+        for (uint i = 0; i < _ops.length; i++) {
+            PackedUserOperation memory op = _ops[i];
+            Transaction memory tx = abi.decode(op.callData, (Transaction));
+
+            // Alignment checks.
+            // These are important in case data will be indexed: we want to make sure that
+            // "wrapped" fields are aligned with the original ones.
+            require(op.sender == address(uint160(tx.from)), "Sender mismatch");
+            require(op.nonce == tx.nonce, "Nonce mismatch");
+            require(op.initCode.length == 0, "Init code not supported");
+            require(uint256(op.accountGasLimits) >> 128 == 0, "Verification gas limit must be 0");
+            require(uint256(op.accountGasLimits) == tx.gasLimit, "Call gas limit mismatch");
```

Comment:
the semantics of verification gas / account gas have changed compared to EIP4337, not sure whether it is a big deal, but advertising it as EIP4337 may not be fully honest